plotly/figure_factory/_facet_grid.py

```diff
@@ -0,0 +1,1073 @@
+            )
+
+    elif facet_row and facet_col:
+        groups_by_facets = list(df.groupby([facet_row, facet_col]))
+        tuple_to_facet_group = {item[0]: item[1] for
+                                item in groups_by_facets}
+
+        row_values = df[facet_row].unique()
+        col_values = df[facet_col].unique()
+        for row_count, x_val in enumerate(row_values):
+            for col_count, y_val in enumerate(col_values):
+                try:
+                    group = tuple_to_facet_group[(x_val, y_val)]
+                except KeyError:
+                    group = pd.DataFrame([[None, None, None]],
+                                         columns=[x, y, color_name])
+
+                if group.values.tolist() != [[None, None, None]]:
+                    trace = graph_objs.Scatter(
+                        x=group[x].tolist(),
+                        y=group[y].tolist(),
+                        mode='markers',
+                        type=trace_type,
+                        marker=dict(
+                            color=df[color_name].tolist(),
+                            size=size,
+                            colorscale=colormap,
+                            showscale=(row_count == 0),
+                            colorbar=dict(x=1.15),
+                            **kwargs_marker
+                        ),
+                        **kwargs
+                    )
+                else:
+                    trace = graph_objs.Scatter(
+                        x=group[x].tolist(),
+                        y=group[y].tolist(),
+                        mode='markers',
+                        type=trace_type,
+                        showlegend=False,
+                        **kwargs
+                    )
+                fig.append_trace(trace, row_count + 1, col_count + 1)
+                if row_count == 0:
+                    label = _return_label(col_values[col_count],
+                                          facet_col_labels, facet_col)
+                    annotations.append(
+                        _annotation_dict(label, col_count + 1, num_of_cols,
+                                         row_col='col', flipped=flipped_cols)
+                        )
+            label = _return_label(row_values[row_count],
+                                  facet_row_labels, facet_row)
+            annotations.append(
+                _annotation_dict(row_values[row_count],
+                                 num_of_rows - row_count, num_of_rows,
+                                 row_col='row', flipped=flipped_rows)
+                )
+
+    # add annotations
+    fig['layout']['annotations'] = annotations
+
+    return fig
+
+
+def _facet_grid(df, x, y, facet_row, facet_col, height, width, num_of_rows,
+                num_of_cols, facet_row_labels, facet_col_labels,
+                trace_type, size, flipped_rows, flipped_cols, show_boxes,
+                kwargs, kwargs_marker):
```

Comment:
this could be a lot DRYer. As far as I can tell, `_facet_grid` is the same as `_facet_grid_color_numerical` which is pretty similar to `_facet_grid_color_categorical`. all of these functions can be merged into one function with a few `if` statements I think. 

For example, 
```
color=df[color_name] if type == 'numerical' else colormap[color_val] if type === 'categorical' else DEFAULT_MARKER_COLOR`
```
or if you prefer:
```
color=(
    df[color_name] if type == 'numerical'
    else colormap[color_val] if type === 'categorical' 
    else DEFAULT_MARKER_COLOR`
)
```

or alternatively, if there are multiple values that depend on the type:

```python
trace = graph_objs.Scatter(
    x=df[x],
    y=df[y],
    mode='markers',
    type=trace_type,
    marker=dict(
        size=size,
        **kwargs_marker
    )
)

if type == 'categorical':
   trace['marker']['color'] = colormap[color_val]
elif type == 'numerical':
   trace['marker']['color'] = df[color_name]
else:
   trace['marker']['color'] = DEFAULT_MARKER_COLOR
```